### PR TITLE
Add Boolean Calculation Command

### LINKED ISSUE #67: Compute Boolean Logical Expression
As a CEG/EE Student, I want to compute a boolean logical expression so that I can familiarise myself with logical computation used in digital circuits.

### LINKED ISSUE #84: Update BooleanTemplate
- To restrict add gate by height
- Check argument length is 1
- Use `isInput` to check if can set input

### PR DESCRIPTION
**Pull request by sevenseasofbri**
Closes #67 and closes #84 
- Broke the code accidentally while doing this, tried to fix it. Seems as though the problem is eradicated for now.
- JUnit testing to be added.
- Updated runtest.bat

### CODE REVIEW

File: src/main/java/seedu/duke/model/gates/Gate.java [hunk 1]
```diff
@@ -1,17 +1,40 @@
...
+        this.secondInput = secondInput;
+    }
+
+    public Gate() {
+        this(-1, -1);
+    }
+
+    public abstract int getOutput() throws DukeException;
 
     public void setInput(int input) {
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/model/gates/Gate.java:29]: Maybe can add JavaDoc here?

File: src/main/java/seedu/duke/model/gates/Gate.java [hunk 2]
```diff
@@ -1,17 +1,40 @@
 package seedu.duke.model.gates;
 
+import seedu.duke.DukeException;
+
 public abstract class Gate {
     protected int output;
     protected int input;
+    protected int secondInput;
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/model/gates/Gate.java:8]: I see, will have to change the class diagrams for this.

File: src/main/java/seedu/duke/model/gates/Gate.java [hunk 3]
```diff
@@ -29,4 +61,34 @@ public int getInput() {
...
+    public boolean isSetSecondInput() {
+        return isSetSecondInput;
+    }
+
+    /**
+     * Returns if both inputs are set or not.
+     *
+     * @return boolean value
+     */
+    public boolean hasSetBothInput() {
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/model/gates/Gate.java]: Why is this necessary? Doesn't seem to be used outside of `checkHasSetBothInput()`.

File: src/main/java/seedu/duke/model/template/BooleanTemplate.java [hunk 4]
```diff
@@ -138,6 +146,24 @@ private String getInputEquation(int index) throws DukeException {
         return equation;
     }
 
+    public int calculateOutput(int idx) throws DukeException {
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/model/template/BooleanTemplate.java]: JavaDoc here?

File: src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java [hunk 5]
```diff
@@ -2,16 +2,18 @@
...
 
     public CalculateBooleanCommand(BooleanTemplate template) {
         super(template);
     }
 
     @Override
     public void execute() throws DukeException {
-
+        ui.printMessage("The output of the current configuration is " + template.calculateOutput(0) + ".");
     }
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java]: Following `CircuitCommand`, what could be done instead is to put this in an overridden `toString()` method:

```suggestion
    }
    
    @Override
    public String toString() {
        return "The output of the current configuration is " + template.calculateOutput(0) + ".\n" + super.toString();
    }
```

Better yet, have a `private int output` variable in the class, and use `.execute()` to set it using `output = template.calculateOutput(0)`. Then you could replace it with `output` in `toString()`.

File: src/main/java/seedu/duke/model/gates/Gate.java (hunk 1 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/gates/Gate.java:29]: Okie sure

File: src/main/java/seedu/duke/model/gates/Gate.java (hunk 2 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/gates/Gate.java:8]: Oof sozz should've informed you

File: src/main/java/seedu/duke/model/gates/Gate.java (hunk 3 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/gates/Gate.java]: You're right! And I was thinking the same, just kept it in case it would be used somewhere else...should I remove it?

File: src/main/java/seedu/duke/model/template/BooleanTemplate.java (hunk 4 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/template/BooleanTemplate.java]: Okay yep

File: src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java (hunk 5 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java]: Good idea! Will do that.

File: src/main/java/seedu/duke/model/gates/Gate.java (hunk 3 above)
Comment (hughjazzman) [reply src/main/java/seedu/duke/model/gates/Gate.java]: Yep can remove.

File: src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java [hunk 6]
```diff
@@ -2,16 +2,24 @@
 
 import seedu.duke.DukeException;
 import seedu.duke.model.template.BooleanTemplate;
+import seedu.duke.ui.Ui;
 
 public class CalculateBooleanCommand extends BooleanCommand {
     public static final String COMMAND_WORD = "calc";
+    public static final Ui ui = new Ui();
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java]: This is no longer necessary, along with the import.
```suggestion
```

File: src/main/java/seedu/duke/model/template/BooleanTemplate.java [hunk 7]
```diff
@@ -138,6 +146,31 @@ private String getInputEquation(int index) throws DukeException {
...
+        if (circuit.isNullAtIndex(idx)) {
+            throw new DukeException("Oops! Nothing set yet.");
+        }
+
+        boolean isNullAtRight = circuit.isNullAtIndex(circuit.getRightIndex(idx));
+        boolean isNullAtLeft = circuit.isNullAtIndex(circuit.getLeftIndex(idx));
+
+        if (!isNullAtLeft) {
+            circuit.getT(idx).setInput(calculateOutput(circuit.getLeftIndex(idx)));
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/model/template/BooleanTemplate.java]: Consider initializing the `Gate` object before these `if` checks? Seems to be `circuit.getT(idx)` is accessed up to 3 times.

File: src/main/java/seedu/duke/model/template/BooleanTemplate.java [hunk 8]
```diff
@@ -138,6 +146,31 @@ private String getInputEquation(int index) throws DukeException {
...
     }
 
+    /**
+     * Recursive function which calculates the output value of the circuit at position passed.
+     *
+     * @param idx int type index of node.
+     * @return int type value output, can be 0 or 1.
+     * @throws DukeException If input values are not set.
+     */
+    public int calculateOutput(int idx) throws DukeException {
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/model/template/BooleanTemplate.java]: This could be overloaded so that `calculateOutput()` will just give `calculateOutput(0)` when called from `CalculateBooleanCommand`. Then this method could also be `private`.

File: src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java (hunk 6 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java]: Noted

File: src/main/java/seedu/duke/model/template/BooleanTemplate.java (hunk 7 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/template/BooleanTemplate.java]: I see, however since this is recursive, the Gate object will continuously be created again and again right...not sure if that's entirely a bad thing

Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/template/BooleanTemplate.java]: Okay I'll change and see

File: src/main/java/seedu/duke/model/template/BooleanTemplate.java (hunk 8 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/template/BooleanTemplate.java]: okay noted

File: src/main/java/seedu/duke/model/template/BooleanTemplate.java (hunk 7 above)
Comment (hughjazzman) [reply src/main/java/seedu/duke/model/template/BooleanTemplate.java]: No new `Gate` object is being created, you are accessing an existing one within `BinaryTree`. Plus you get much more clarity.

Comment (hughjazzman) [reply src/main/java/seedu/duke/model/template/BooleanTemplate.java]: On that note, you could also have a variable for the left and right index.

File: src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java [hunk 9]
```diff
@@ -5,13 +5,19 @@
...
     }
 
     @Override
     public void execute() throws DukeException {
+        output = template.calculateOutput();
+    }
 
+    @Override
+    public String toString() {
+        return "The output of the current configuration is " + output + ".";
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java]:
```suggestion
        return super.toString() + System.lineSeparator() + "The output of the current configuration is " + output + ".";
```

Comment (sevenseasofbri) [reply src/main/java/seedu/duke/logic/commands/gates/CalculateBooleanCommand.java]: OH NO i put it after lmao

File: docs/UserGuide.md [hunk 10]
```diff
@@ -341,7 +341,7 @@ C = F OR G
 F = 1
 G = 0
 
-The output of the above configuration is 0.
+The output of the current configuration is 0.
```
Comment (hughjazzman) [inline docs/UserGuide.md]: Following the implementation this should above the template.

Comment (sevenseasofbri) [reply docs/UserGuide.md]: okie

**Review by hughjazzman — APPROVED** [review]
LGTM
